### DISCUSSION TITLE
How to extend the official firmware from a custom ROS package?

### DISCUSSION THREAD

**Discussion by antalpeter1** [USC-ACTLab/crazyswarm#545] [Q&A]
We have created a custom ROS package based on the [crazyswarm documentation](https://crazyswarm.readthedocs.io/en/latest/howto/howto.html#option-2-custom-ros-package), and it works successfully for our launch files and Python scripts. However, we can not find out how to link our custom firmware C code from the custom ROS package to the official firmware, build them together and upload to the Crazyflie. We would like to manage this without any modifications on the official firmware code.
Could you please help us?
Thanks in advance,
Peter

**Comment by whoenig** [comment]
Could you clarify what you mean by "link our custom firmware C code [...] to the official firmware" and "without any modifications on the official firmware code"? If you have custom C code, this automatically changes the firmware image. Please also clarify if the problem is in the building firmware step or the firmware flashing step.

**Reply by antalpeter1** [reply]
We have written a new low level controller and we would like to test it on the Crazyflie (by changing the firmware image, of course). However, we do not want to modify the source code in the `crazyswarm/crazyflie-firmware`  directory, but add source code from an outside directory, e.g. `mycrazypack/myfirmware/mycontroller.c` (this is what I meant by linking). The problem is both with building and flashing the firmware, we do not know any of the two in this case (where to put the Makefile and what to include in it).

**Comment by whoenig** [comment]
You can simply keep the firmware external and compile it using make. If you want to flash it with chooser.py, you can specify the path to your firmware images, see https://github.com/USC-ACTLab/crazyswarm/blob/4d6ca47b085227fbc893479894001d1c7ceab5cc/ros_ws/src/crazyswarm/scripts/chooser.py#L20-L30.

The firmware in crazyswarm/crazyflie-firmware is only used to build Python bindings for the SIL simulation. These bindings will be part of the official firmware repository soon (see e.g., https://github.com/bitcraze/crazyflie-firmware/pull/886).
